mono/utils/mono-log-common.c

```diff
@@ -0,0 +1,143 @@
+/*
+ * mono-log-common.c: Platform-independent interface to the logger
+ *
+ * This module contains the POSIX syslog logger interface
+ *
+ * Author:
+ *    Neale Ferguson <neale@sinenomine.net>
+ *
+ */
+#include <config.h>
+
+#ifdef HAVE_UNISTD_H
+#include <unistd.h>
+#endif
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <ctype.h>
+#include <string.h>
+#include <glib.h>
+#include <errno.h>
+#include <time.h>
+#ifndef HOST_WIN32
+#include <sys/time.h>
+#endif
+#include "mono-logger.h"
+
+static FILE *logFile = NULL;
+static void *logUserData = NULL;
+
+/**
+ * mapSyslogLevel:
+ * 	
+ * 	@level - GLogLevelFlags value
+ * 	@returns The equivalent character identifier
+ */
+static inline char 
+mapLogFileLevel(GLogLevelFlags level) 
+{
+	if (level & G_LOG_LEVEL_ERROR)
+		return ('E');
+	if (level & G_LOG_LEVEL_CRITICAL)
+		return ('C');
+	if (level & G_LOG_LEVEL_WARNING)
+		return ('W');
+	if (level & G_LOG_LEVEL_MESSAGE)
+		return ('N');
+	if (level & G_LOG_LEVEL_INFO)
+		return ('I');
+	if (level & G_LOG_LEVEL_DEBUG)
+		return ('D');
+	return ('I');
+}
+
+/**
+ * mono_log_open_logfile
+ * 	
+ *	Open the logfile. If the path is not specified default to stdout. If the
+ *	open fails issue a warning and use stdout as the log file destination.
+ *
+ * 	@path - Path for log file
+ * 	@userData - Not used
+ */
+void
+mono_log_open_logfile(const char *path, void *userData)
+{
+	if (path == NULL) {
+		logFile = stdout;
+	} else {
+		logFile = fopen(path, "w");
+		if (logFile == NULL) {
+			g_warning("opening of log file %s failed with %s - defaulting to stdout", 
+				  path, strerror(errno));
+		}
+	}
+	logUserData = userData;
+}
+
+/**
+ * mono_log_write_logfile
+ * 	
+ * 	Write data to the log file.
+ *
+ * 	@domain - Identifier string
+ * 	@level - Logging level flags
+ * 	@format - Printf format string
+ * 	@vargs - Variable argument list
+ */
+void
+mono_log_write_logfile(const char *domain, GLogLevelFlags level, mono_bool hdr, const char *format, va_list args)
+{
+	time_t t;
+	char logTime[80],	
+	     logMessage[512];
+	pid_t pid;
+	int iLog = 0;
+	size_t nLog;
+
+	if (logFile == NULL)
+		logFile = stdout;
+
+	if (hdr) {
+#ifndef HOST_WIN32
+		struct tm tod;
+		time(&t);
+		localtime_r(&t, &tod);
+		pid = getpid();
+		strftime(logTime, sizeof(logTime), "%F %T", &tod);
+#else
+		struct tm *tod;
+		time(&t);
+		tod = localtime(&t);
+		pid = _getpid();
+		strftime(logTime, sizeof(logTime), "%F %T", tod);
+#endif
+		iLog = snprintf(logMessage, sizeof(logMessage), "%s level[%c] mono[%d]: ",
```

Comment:
Ah, they behave differently on UNIX where truncation is indicated differently. Will fix.

From: Johan Lorensson +ADw-notifications+AEA-github.com+ADw-mailto:notifications+AEA-github.com+AD4APg-
Reply-To: mono/mono +ADw-reply++0004b6fab10257568976a73ab2833a54254425eaa2c73db392cf000000011368f2b192a170ce03e9fYAQA-reply.github.com+ADw-mailto:reply++0004b6fab10257568976a73ab2833a54254425eaa2c73db392cf000000011368f2b192a170ce03e9fYAQA-reply.github.com+AD4APg-
Date: Friday, June 3, 2016 at 3:24 AM
To: mono/mono +ADw-mono+AEA-noreply.github.com+ADw-mailto:mono+AEA-noreply.github.com+AD4APg-
Cc: Neale Ferguson +ADw-neale+AEA-sinenomine.net+ADw-mailto:neale+AEA-sinenomine.net+AD4APg-, Mention +ADw-mention+AEA-noreply.github.com+ADw-mailto:mention+AEA-noreply.github.com+AD4APg-
Subject: Re: +AFs-mono/mono+AF0- Enhance log tracing (+ACM-2734)


In mono/utils/mono-log-common.c+ADw-https://github.com/mono/mono/pull/2734+ACM-discussion+AF8-r65665806+AD4-:

+AD4-
+AD4-       if (hdr) +AHs-
+AD4- +ACM-ifndef HOST+AF8-WIN32
+AD4-               struct tm tod+ADs-
+AD4-               time(+ACY-t)+ADs-
+AD4-               localtime+AF8-r(+ACY-t, +ACY-tod)+ADs-
+AD4-               pid +AD0- getpid()+ADs-
+AD4-               strftime(logTime, sizeof(logTime), +ACIAJQ-F +ACU-T+ACI-, +ACY-tod)+ADs-
+AD4- +ACM-else
+AD4-               struct tm +ACo-tod+ADs-
+AD4-               time(+ACY-t)+ADs-
+AD4-               tod +AD0- localtime(+ACY-t)+ADs-
+AD4-               pid +AD0- +AF8-getpid()+ADs-
+AD4-               strftime(logTime, sizeof(logTime), +ACIAJQ-F +ACU-T+ACI-, tod)+ADs-
+AD4- +ACM-endif
+AD4-               iLog +AD0- snprintf(logMessage, sizeof(logMessage), +ACIAJQ-s level+AFsAJQ-c+AF0- mono+AFsAJQ-d+AF0-: +ACI-,


snprintf can return negative value on failures.

+IBQ-
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub+ADw-https://github.com/mono/mono/pull/2734/files/583d1a2580a5fc61d0002920c17c23fa71375c3d+ACM-r65665806+AD4-, or mute the thread+ADw-https://github.com/notifications/unsubscribe/AAS2-lO8OnEiQmfQ5hUxJvt56L379LZUks5qH9axgaJpZM4HsGtY+AD4-.